### PR TITLE
SNOW-625820 Schema mapping

### PR DESCRIPTION
**Pull request by sfc-gh-zefan**
Added schema mapping logic. Data will be loaded into respective column through name matching of the field.
Added logic to add the 'RECORD_METADATA' column if not present.
Added a config to enable the above features.
Added tests.

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 1]
```diff
@@ -57,6 +64,8 @@ public class RecordService extends Logging {
   private static final String KEY_SCHEMA_ID = "key_schema_id";
   static final String HEADERS = "headers";
 
+  private boolean schematizationEnable = false;
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:67]: `enableSchematization`

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 2]
```diff
@@ -87,6 +96,17 @@ public void setMetadataConfig(SnowflakeMetadataConfig metadataConfigIn) {
     metadataConfig = metadataConfigIn;
   }
 
+  public boolean setSchematizationEnable(final Map<String, String> connectorConfig) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:99]: Java docs for public methods.

File: test/test_verify.py [hunk 3]
```diff
@@ -437,7 +448,7 @@ def runTestSet(driver, testSet, nameSalt, pressure):
 
     testStringJsonProxy = TestStringJsonProxy(driver, nameSalt)
 
-    # Run this test on both confluent and apache kafka
+    # # Run this test on both confluent and apache kafka
```
Comment (sfc-gh-japatel) [inline test/test_verify.py:451]: do you need this change?

File: test/test_suit/test_schema_mapping.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+from test_suit.test_utils import RetryableError, NonRetryableError
+import json
+
+
+class TestSchemaMapping:
```
Comment (sfc-gh-japatel) [inline test/test_suit/test_schema_mapping.py:5]: A small description here would be great.

**Comment by sfc-gh-japatel** [comment]
In the description you have `Added logic to add the 'RECORD_CONTENT' column if not present.`
Do you mean record_metadata?

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 5]
```diff
@@ -326,6 +330,50 @@ public boolean isTableCompatible(final String tableName) {
...
+      }
+    } catch (SQLException e) {
+      logError("table {} doesn't exist", tableName);
+      throw SnowflakeErrors.ERROR_2014.getException("table name: " + tableName);
+    }
+    try {
+      if (!isVariant) {
+        String metaQuery;
+        if (!hasMeta) {
+          metaQuery = "alter table identifier(?) add RECORD_METADATA VARIANT";
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:364]: Do we need to make any changes in the docs when we publish about schema mapping, about some specific privilege/permissions?

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 6]
```diff
@@ -177,7 +207,20 @@ public Map<String, Object> getProcessedRecordForStreamingIngest(SinkRecord recor
...
+        if (enableSchematization) {
+          Iterator<String> fieldNames = node.fieldNames();
+          while (fieldNames.hasNext()) {
+            String fieldName = fieldNames.next();
+            JsonNode fieldNode = node.get(fieldName);
+            String filedContent =
+                fieldNode.isTextual()
+                    ? fieldNode.textValue()
+                    : MAPPER.writeValueAsString(fieldNode);
+            streamingIngestRow.put(fieldName, filedContent);
```
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:219]: Also, I am assuming we only care about one level deep jsonNode? all rest of them are going into a variant column?

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 7]
```diff
@@ -177,7 +207,20 @@ public Map<String, Object> getProcessedRecordForStreamingIngest(SinkRecord recor
     final Map<String, Object> streamingIngestRow = new HashMap<>();
     for (JsonNode node : row.content.getData()) {
       try {
-        streamingIngestRow.put(TABLE_COLUMN_CONTENT, MAPPER.writeValueAsString(node));
+        if (enableSchematization) {
+          Iterator<String> fieldNames = node.fieldNames();
+          while (fieldNames.hasNext()) {
+            String fieldName = fieldNames.next();
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:213]: nit: more like a columnName?

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 8]
```diff
@@ -177,7 +207,20 @@ public Map<String, Object> getProcessedRecordForStreamingIngest(SinkRecord recor
...
+            JsonNode fieldNode = node.get(fieldName);
+            String filedContent =
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:215]: nit: more like a `columnValue`

**Review by sfc-gh-japatel — CHANGES_REQUESTED** [review]
I think overall looks great, mostly cosmetic changes. Will wait for one more pass. 

Also, tests looks great, thanks for adding them!

File: test/test_suit/test_schema_mapping.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    def __init__(self, driver, nameSalt):
+        self.driver = driver
+        self.fileName = "travis_correct_schema_mapping"
+        self.topic = self.fileName + nameSalt
+
+        self.driver.snowflake_conn.cursor().execute(
+            "Create or replace table {} (PERFORMANCE_STRING STRING, PERFORMANCE_CHAR CHAR, PERFORMANCE_HEX BINARY, RATING_INT NUMBER, RATING_DOUBLE DOUBLE, APPROVAL BOOLEAN, APPROVAL_DATE DATE, APPROVAL_TIME TIME)".format(self.topic))
```
Comment (sfc-gh-japatel) [inline test/test_suit/test_schema_mapping.py:12]: can you squeeze in a variant here too??

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 10]
```diff
@@ -177,7 +207,20 @@ public Map<String, Object> getProcessedRecordForStreamingIngest(SinkRecord recor
...
+                fieldNode.isTextual()
+                    ? fieldNode.textValue()
+                    : MAPPER.writeValueAsString(fieldNode);
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:218]: can you add a UT which will test this line? (I assume this is when a field itself is a json string)

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sfc-gh-zefan) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:219]: will do; Yep

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java (hunk 5 above)
Comment (sfc-gh-zefan) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:364]: The assumption is that the user themself needs to create the table with the correct schema. They will need the privilege to alter the table if the metadata column is added by us. In that case they need one of 'INSERT', 'ALL' or 'OWNERSHIP'.

File: test/test_suit/test_schema_mapping.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
+from test_suit.test_utils import RetryableError, NonRetryableError
+import json
+
+# test if each type of data fit into the right column with the right type
+# also test if the metadata column is automatically added
+class TestSchemaMapping:
```
Comment (sfc-gh-tzhang) [inline test/test_suit/test_schema_mapping.py:6]: if the kafka message has two columns c1 and C1, will we be able to mapped to the current column in the table?

Comment (sfc-gh-zefan) [reply test/test_suit/test_schema_mapping.py:6]: I believe they will be mapped into the same column C1 with the current behavior. I think we could add an option in the sf config to allow uppercase letter, and we can add the quotation marks in the connector.

Comment (sfc-gh-tzhang) [reply test/test_suit/test_schema_mapping.py:6]: Are you saying the second value will override the first value?

Comment (sfc-gh-zefan) [reply test/test_suit/test_schema_mapping.py:6]: I suppose so. I think this happens in the SDK.

Comment (sfc-gh-tzhang) [reply test/test_suit/test_schema_mapping.py:6]: Could we add a verification step in KC to check whether there're two columns with the same name if ignoring cases, and fail saying we can't supported?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 12]
```diff
@@ -368,6 +370,10 @@ private boolean isContentBroken(final Object content) {
     return content != null && ((SnowflakeRecordContent) content).isBroken();
   }
 
+  /**
+   * TODO: SNOW-630885 - When schematization is enabled, we should create the map directly from the
+   * SinkRecord instead of first turning it into json
+   */
```
[the remark this answers is not in the document]
Comment (sfc-gh-zefan) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:376]: Indeed there's no change here, but I believe this is where the change should be made. A function similar to the current handleNativeRecord should be added to produce a map return value.

File: test/test_suit/test_schema_mapping.py (hunk 11 above)
Comment (sfc-gh-zefan) [reply test/test_suit/test_schema_mapping.py:6]: Just added the check and tested it. I could add a UT if you think it's necessary

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 13]
```diff
@@ -186,6 +253,11 @@ public Map<String, Object> getProcessedRecordForStreamingIngest(SinkRecord recor
         e.printStackTrace();
       }
     }
+
+    if (hasCaseInsensitiveDuplicates(streamingIngestRow)) {
+      throw SnowflakeErrors.ERROR_0025.getException();
+    }
+
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:260]: discussed offline, we don't need to do this anymore, we could have a function called `formatColumnName` which convert everything to uppercase if no double quotes, please take a look at how [ingest SDK](https://github.com/snowflakedb/snowflake-ingest-java/blob/83008cd9259fb18518f3167261aacdc664bdb992/src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java#L617) is handling it

Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:260]: And could you add a test on double quotes column name?

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM other than my comments, thanks!

File: src/test/java/com/snowflake/kafka/connector/records/RecordContentTest.java [hunk 14]
```diff
@@ -230,4 +231,44 @@ public void testConvertToJsonReadOnlyByteBuffer() {
     Schema schema = SchemaBuilder.bytes().build();
     assert RecordService.convertToJson(schema, buffer).toString().equals(expected);
   }
+
+  @Test
+  public void testSchematizationStringField() {
+    RecordService service = new RecordService();
+    SnowflakeJsonConverter jsonConverter = new SnowflakeJsonConverter();
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/records/RecordContentTest.java:238]: can you instead use JsonConverter in both the tests below? 

We are not using JsonConverter for `SNOWPIPE_STREAMING`

File: test/test_suit/test_schema_mapping.py (hunk 4 above)
Comment (sfc-gh-japatel) [reply test/test_suit/test_schema_mapping.py:6]: Thanks @sfc-gh-zefan

**Review by sfc-gh-japatel — APPROVED** [review]
apart from one change in test, lgtm! Approving since your test has jsonconverter so I am assuming its working! 

Great stuff @sfc-gh-zefan

**Comment by sfc-gh-japatel** [comment]
My only concern is if it is okay to not push it to master? 

You can push it to some `schema-mapping-master` branch.. 

Only reason to do this is we will be releasing new versions and we will have to pick our commits. Once we are ready for PrPr, we can bring the commits from `schema-mapping-master` to `master`

Thoughts @sfc-gh-tzhang @sfc-gh-zefan ?

**Pull request merged**